datafusion/physical-plan/src/topk/mod.rs

```diff
@@ -319,19 +342,75 @@ impl TopK {
     /// (a > 2 OR (a = 2 AND b < 3))
     /// ```
     fn update_filter(&mut self) -> Result<()> {
-        let Some(filter) = &self.filter else {
+        // If the heap doesn't have k elements yet, we can't create thresholds
+        let Some(max_row) = self.heap.max() else {
             return Ok(());
         };
-        let Some(thresholds) = self.heap.get_threshold_values(&self.expr)? else {
-            return Ok(());
+
+        let new_threshold_row = &max_row.row;
+
+        // Extract filter expression reference before entering critical section
+        let filter_expr = Arc::clone(&self.filter.expr);
+
+        // Check if we need to update and do both threshold and filter update atomically
+        {
+            let mut threshold_guard = self.filter.threshold_row.write();
+            if let Some(current_row) = threshold_guard.as_ref() {
+                match current_row.as_slice().cmp(new_threshold_row) {
+                    Ordering::Greater => {
+                        // new < current, so new threshold is more selective
```

Comment:
I think this was wrong before @adriangb  - in the heap lower means more selective

Reply:
The current solution seems roughly on par with main, I don't observe a speedup.

Reply:
I guess we should try changing it to the global heap thing - write updates to the shared/global heap and update the filter based on global heap.